src/i2c.rs

```diff
@@ -101,6 +101,58 @@
 
 use crate::private;
 
+/// I2C error
+pub trait Error: core::fmt::Debug {
+    /// Convert error to a generic I2C error kind
+    ///
+    /// By using this method, I2C errors freely defined by HAL implementations
+    /// can be converted to a set of generic I2C errors upon which generic
+    /// code can act.
+    fn kind(&self) -> ErrorKind;
+}
+
+/// I2C error kind
+///
+/// This represents a common set of I2C operation errors. HAL implementations are
+/// free to define more specific or additional error types. However, by providing
+/// a mapping to these common I2C errors, generic code can still react to them.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// An unspecific bus error occurred
+    Bus,
+    /// The arbitration was lost, e.g. electrical problems with the clock signal
+    ArbitrationLoss,
+    /// A bus operation was not acknowledged, e.g. due to the addressed device not being available on
+    /// the bus or the device not being ready to process requests at the moment
+    NoAcknowledge,
```

Comment:
It would be useful to distinguish between "NAK on address" and "NAK on data". First may indicate the slave is not there (or busy), the second indicates the slave is deinitely there, but isn't accepting (more) data. 

nRFs and STM32s can distinguish them, probably others too.

Reply:
I'm only aware of a "NACK on data" for SMBus and haven't seen that supported for I2C yet. Which STM32 supports that if I may ask?

Reply:
I have seen a few micro controllers support this (though those that come to mind don't have publicly available datasheets). I'm not sure how useful this information will be in practice as I cannot think of any time anything but the I2C peripheral driver itself would care about this distinction.

Reply:
If a microcontroller doesn't support distinguishing the two, mapping a NACK to one of those two variants would be difficult.

Reply:
That is a good point. The two solutions I could see here would either to just keep one `NoAcknowledge` variant or to have `NoAcknowledgeAddress` and `NoAcknowledgeData` variants and specify that when a driver can't differentiate between the two to default to a specific one.

I think I would prefer the second option, but I would not be against the first.

Reply:
I propose doing this:

- NoAcknowledge: NAK after addr or data byte, the hardware is incapable of distinguishing them.
- NoAcknowledgeAddress: NAK after address byte
- NoAcknowledgeData: NAK after data byte

Reply:
That seems fine by me. There would just need to be documentation detailing this, of course.

Reply:
I replaced `NoAcknowledge` with `NoAcknowledgeAddress` and `NoAcknowledgeData`. Since the `enum` is non-exhaustive, we can add `NoAcknowledge` later if an implemementation cannot distinguish between the two.

Reply:
Can you imagine a generic driver handling differently NAK address and NAK data? Because if that's not the case, better to do the contrary: only keep NoAcknowledge for now, and we can add the address and data variant later, no?

Reply:
Yes, depending on the device, NAK address may mean the device is not there at all so bubble up the error to the caller. OTOH, NAK data may mean "I'm busy right now, but you can try again later" so the driver can retry at a later time or return something like `nb::WouldBlock` or so.

Reply:
OK, make sense.

Reply:
I think only having `NoAcknowledgeAddress` and `NoAcknowledgeData` without an defined way to handle drivers that cannot differentiate between the two is problematic. If we don't want to add a third `NoAcknowledge`, I would ask that we add documentation saying which one to use by default. Adding `NoAcknowledge` down the line with the current design might cause users to miss the new `NoAcknowledge`, even if adding that variant was considered only a minor change.

---

This may be unacceptable, but perhaps we should instead have `NoAcknowledge` be a tuple enum that specifies if it was on the address or data. e.g.

```rust
#[non_exhaustive]
pub enum ErrorKind {
        /// A bus operation was not acknowledged, e.g. due to the addressed device not being available on
    /// the bus or the device not being ready to process requests at the moment
    NoAcknowledge(NoAcknowledgeSource),
}

enum NoAcknowledgeSource {
    /// The device did not acknowledge its address. The device may be missing.
    Address,
    /// The device did not acknowledge the data. It may not be ready to process requests at the moment.
    Data,
    /// The driver does not know if the no acknowledge was from the address or the data.
    Unknown,
}
```

This would allow users to catch all NACKs if they wanted to or get more information if they wanted that.

Reply:
Since we have reached an agreement on this long-standing issue, I would like to merge it as it is.
If you are interested, would you open a separate PR or issue with this proposal?

Reply:
Yeah, I will submit an follow-on PR.